automata/fa/nfa.py

```diff
@@ -1024,3 +985,65 @@ def add_any_transition(start_state_dict, end_state):
             initial_state=(0, 0),
             final_states=final_states,
         )
+
+    def iter_states(self):
+        return iter(self.states)
+
+    def iter_transitions(self):
+        return (
+            (from_, to_, symbol)
+            for from_, lookup in self.transitions.items()
+            for symbol, to_lookup in lookup.items()
+            for to_ in to_lookup
+        )
+
+    def is_accepting(self, state):
+        return state in self.final_states
+
+    def is_initial(self, state):
+        return state == self.initial_state
+
+    def get_edge_name(self, symbol):
+        return "ε" if symbol == "" else str(symbol)
+
+    def _get_input_path(self, input_str):
+        """Calculate the path taken by input."""
+
+        visiting = set()
+
+        def gen_paths_for(state, step):
```

Comment:
@eliotwrobson I might have overthought this one! The reason for the two mutually recursive functions is to find a preferably accepted path with the fewest jumps and longest input read if the input was not accepted. This can be simplified with some hit to the path found (trying to find the first path for example). The `read_input_stepwise` method does not share a lot of logic with these two functions. We don't need to keep track of the full state of nfa when performing a pathfinding algorithm (DFS). It would have been better if `read_input_stepwise` were to take `starting_state` as an argument instead of using `initial_state` as the starting state all the time. And even then we still would have needed all the transitions used in each jump which we don't get.

Reply:
@khoda81 that makes sense. Not a huge deal if these two functions can't share any logic (as you said, `read_input_stepwise` only really cares about the current set of states not the exact transitions taken), but it would be nice to implement this function iteratively. I think finding the simplest path should be doable with BFS?

Also, be sure to test these functions with an NFA that has a cycle of lambda transitions, I think it might behave strangely in that situation.

Reply:
Resolving since this function was rewritten.